dbms/src/IO/Compression/CompressionCodecLightweight_Integer.cpp

```diff
@@ -0,0 +1,359 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/Exception.h>
+#include <IO/Compression/CompressionCodecLightweight.h>
+#include <IO/Compression/CompressionSettings.h>
+#include <IO/Compression/EncodingUtil.h>
+#include <lz4.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int CANNOT_COMPRESS;
+extern const int CANNOT_DECOMPRESS;
+} // namespace ErrorCodes
+
+String CompressionCodecLightweight::IntegerCompressContext::toDebugString() const
+{
+    return fmt::format(
+        "lz4: {}, lightweight: {}, constant_delta: {}, delta_for: {}, rle: {}, lz4 {} -> {}, lightweight {} -> {}",
+        lz4_counter,
+        lw_counter,
+        constant_delta_counter,
+        delta_for_counter,
+        rle_counter,
+        lz4_uncompressed_size,
+        lz4_compressed_size,
+        lw_uncompressed_size,
+        lw_compressed_size);
+}
+
+void CompressionCodecLightweight::IntegerCompressContext::update(size_t uncompressed_size, size_t compressed_size)
+{
+    if (mode == IntegerMode::LZ4)
+    {
+        lz4_uncompressed_size += uncompressed_size;
+        lz4_compressed_size += compressed_size;
+        ++lz4_counter;
+    }
+    else
+    {
+        lw_uncompressed_size += uncompressed_size;
+        lw_compressed_size += compressed_size;
+        ++lw_counter;
+    }
+    if (mode == IntegerMode::CONSTANT_DELTA)
+        ++constant_delta_counter;
+    if (mode == IntegerMode::DELTA_FOR)
+        ++delta_for_counter;
+    if (mode == IntegerMode::RunLength)
+        ++rle_counter;
+}
+
+bool CompressionCodecLightweight::IntegerCompressContext::needAnalyze() const
+{
+    // lightweight codec is never used, do not analyze anymore
+    if (lz4_counter > COUNT_THRESHOLD && lw_counter == 0)
+        return false;
+    // if lz4 is used more than COUNT_THRESHOLD times and the compression ratio is better than lightweight codec, do not analyze anymore
+    if (lz4_counter > COUNT_THRESHOLD
+        && lz4_uncompressed_size / lz4_compressed_size > lw_uncompressed_size / lw_compressed_size)
+        return false;
+    return true;
+}
+
+bool CompressionCodecLightweight::IntegerCompressContext::needAnalyzeDelta() const
+{
+    return lw_counter <= COUNT_THRESHOLD || constant_delta_counter != 0 || delta_for_counter != 0;
+}
+
+bool CompressionCodecLightweight::IntegerCompressContext::needAnalyzeRunLength() const
+{
+    return lw_counter <= COUNT_THRESHOLD || rle_counter != 0;
+}
+
+template <typename T>
+void CompressionCodecLightweight::IntegerCompressContext::analyze(std::span<const T> & values, IntegerState<T> & state)
+{
+    if (values.empty())
+    {
+        mode = IntegerMode::Invalid;
+        return;
+    }
```

Comment:
Seems if we want to apply lightweight compression with `source_size == 0`, this will lead to exception

Reply:
`source_size` always greater than 0.